### ISSUE TITLE
Issue compiling metric function

### ISSUE THREAD

**Issue by piyushjo15** [lmcinnes/pynndescent#117]
Hi,

I am trying to use NNDescent and PyNNDescentTransformer function using a custom metric function. Actually, it is not properly custom, as I am trying to use 'jensenshannon' metric from scipy.spatial.distance. I am trying to use this metric, that is not available in pynndescent. Below if my code that I am trying to run. 

```
import time
import numpy as np
import numba
import pynndescent as nd
import scipy.spatial.distance as sc

#input matrix, 2000x20 matrix
rd_nmf = np.genfromtxt("rd_nmf2.csv", delimiter = ',')

@numba.jit(forceobj=True)
def jsd(x,y):
	return sc.jensenshannon(x,y)

start = time.time()
Y2 = nd.PyNNDescentTransformer(metric=jsd)
td2 = Y2.fit_transform(rd_nmf2)
print("--- %s seconds ---" % (time.time() - start))

```
Now, I have checked that my 'jsd' function does give the required value for two arrays x and y, just like running sc.jensenshannon(x,y), so the function itself is not giving error. Also I put the 'forceobj=True', which also prevented the compiling from giving any error. When I try to run the 'fit_trasnform' part, I get errors. Now I am not so much used to python and some of the errors I can't understand and I did google them before posting this here, but I couldn't solve it. Could you please help?

**Comment by lmcinnes** [comment]
The difficulty here is that you need to be able to compile the function in non-object mode, and unfortunately the scipy function if an object that can't be compiled in this case. Essentially the problem is that you need an implementation in terms of more basic numpy primitives that numba can compile down to machine code.

It looks like the scipy implementation uses a combination of pure python, and some Cython (presumably vectorized). The python code is here: https://github.com/scipy/scipy/blob/v1.6.0/scipy/spatial/distance.py#L1244-L1301 and the Cython code to compute relative_entropy is here: https://github.com/scipy/scipy/blob/v1.6.0/scipy/spatial/distance.py#L1244-L1301

We can stitch that together ourselves in numba compilable code roughly as follows:

```python
@numba.vectorize(fastmath=True)
def relative_entropy(x, y):
    if np.isnan(x) or np.isnan(y):
        return np.nan
    elif x > 0 and y > 0:
        return x * np.log(x / y)
    elif x == 0 and y >= 0:
        return 0.0
    else:
        return np.inf

@numba.njit(
    nogil=True, 
    fastmath=True,
)
def jensen_shannon(p, q, base=None):
    p = p / np.sum(p, axis=0)
    q = q / np.sum(q, axis=0)
    m = (p + q) / 2.0
    left = relative_entropy(p, m)
    right = relative_entropy(q, m)
    js = np.sum(left, axis=0) + np.sum(right, axis=0)
    if base is not None:
        js /= np.log(base)
    return np.sqrt(js / 2.0)
```

This  should now, work, and will produce the same results as the scipy code (since it is really largely duplicating it). Depending on how badly you want this to run fast you can go a little further and use

```python
@numba.njit(fastmath=True, nogil=True)
def rel_entr(x, y):
    if np.isnan(x) or np.isnan(y):
        return np.nan
    elif x > 0 and y > 0:
        return x * np.log(x / y)
    elif x == 0 and y >= 0:
        return 0.0
    else:
        return np.inf

@numba.njit(
    [
        "f4(f4[::1],f4[::1])",
        numba.types.float32(
            numba.types.Array(numba.types.float32, 1, "C", readonly=True),
            numba.types.Array(numba.types.float32, 1, "C", readonly=True),
        ),
    ],
    nogil=True, 
    fastmath=True,
    locals={
        "pnorm": numba.float32,
        "qnorm": numba.float32,
        "mnorm": numba.float32,
        "pi": numba.float32,
        "qi": numba.float32,
        "m": numba.float32,
        "result": numba.float32,
    }
)
def jensen_shannon_fast(p, q):
    result = 0.0
    pnorm = 0.0
    qnorm = 0.0
    for i in range(p.shape[0]):
        pnorm += p[i]
        qnorm += q[i]
    mnorm = pnorm + qnorm
    for i in range(p.shape[0]):
        pi = p[i] / pnorm
        qi = q[i] / qnorm
        m = (pi + qi) / mnorm
        result += rel_entr(pi, m)
        result += rel_entr(qi, m)
    
    return np.sqrt(result / 2.0)
```
which gave me 2/3 of the the runtime in microbenchmarking, and will still be acceptable to pynndescent (although it may give slightly less accurate answers -- we are forcing everything to work in float32, which may not provide enough accuracy for you).

**Comment by piyushjo15** [comment]
Amazing!! Thanks a lot for this! This definitely speed up the code. One more thing I wanted to ask, when I use transform to obtain the graph, I want to have all the connected nodes as 1 and unconnected nodes as 0. I think right now the result is weighted, so there are non-negative numbers for connected neighbors. I couldn't find an option to make the graph equally weighted. Thanks again for the code!

**Comment by lmcinnes** [comment]
If you are using the PyNNDescentTransformer transform then the result should be a sparse matrix. In that case you just need to convert all the non-zero entries to ones. The easiest way to do that is to just check for inequality to zero, which works well with scipy.sparse matrices. In other words:

```python
matrix = index.transform(data)
binary_matrix = (matrix != 0)
```

and ``binary_matrix`` should be what you want.

**Comment by piyushjo15** [comment]
Hi,

I just wanted to share this error that I am getting from your fast method.
`No matching definition for argument type(s) array(float64, 1d, C), array(float64, 1d, C)`

Probably something is wrong with my argument passing. When I removed the 
```
 [
        "f4(f4[::1],f4[::1])",
        numba.types.float32(
            numba.types.Array(numba.types.float32, 1, "C", readonly=True),
            numba.types.Array(numba.types.float32, 1, "C", readonly=True),
        ),
    ],
```

I was able to run the function. But the return value is not the same as your normal function which gives the same value as scipy jensenshannon function. I am happy with the slow function that you wrote, but just wanted to let you know of this issue. I will close this thread now. Thanks a lot again!
